### DISCUSSION TITLE
How can I restrict user registration to only be allowed to super users?

### DISCUSSION THREAD

**Discussion by augusto-herrmann** [fastapi-users/fastapi-users#518] [Q&A] [answered]
The documentation explains how you can restrict an API method to be [allowed only for superusers](https://frankie567.github.io/fastapi-users/usage/dependency-callables/#get-the-current-active-superuser). However, the user register is not defined by the API code, but by `fastapi_users` itself:

```python
app.include_router(
    fastapi_users.get_register_router(on_after_register),
    prefix="/auth",
    tags=["auth"]
)
```

How can I apply a restriction here that only authenticated super users can use this method?

**Comment by frankie567** [comment, marked as answer]
You can define the dependency at a router level (https://fastapi.tiangolo.com/tutorial/bigger-applications/#include-an-apirouter-with-a-custom-prefix-tags-responses-and-dependencies):

```py
app.include_router(
    fastapi_users.get_register_router(on_after_register),
    prefix="/auth",
    tags=["auth"],
    dependencies=[Depends(fastapi_users.current_user(active=True, superuser=True))],
)
```

All the sub-routes will then call this dependency.

Another solution is to write your own user creation route, that will be more adapted maybe for a superuser, and not include the register router.

**Reply by augusto-herrmann** [reply]
That looks good, thanks! I will try doing it that way.

**Comment by rbracco** [comment]
I have a related question so I figured I'd ask here rather than start a new topic. I would like to implement an invite code as part of the auth, so that only users with a valid invite code are able to register a new account. I have the invite validation functionality in a function called `validate_invite_code` but I'm having trouble figuring out how to connect it to the auth router. 

Options seem to be 
1. Passing `dependencies=[Depends(validate_invite_code(invite_code_needed_here))]` to the auth router but I'm not sure how to get the invite code from the request body and pass it as an arg to the function
2. Write my own user creation route and then call the invite validation there. I'm also unsure about how to create the user creation route myself and I was unable to find anything in the docs. Any guidance on the easiest way to implement invites would be much appreciated. Thank you!

**Reply by frankie567** [reply]
Hi @rbracco!

It depends on what you wish to want to do with this invite code.

If it's just checking that the invitation exists in database and then proceed with standard registration, you can go for the solution 1. The dependency function could look like this:

```py
async def validate_invite_code(code: str = Query(...)):  # FastAPI will automatically inject the query parameter named code or raise 422 if not present
  invitation = await db.get_invitation(code)
  if not invitation:
    raise HTTPException(status_code=403, detail="This invitation doesn't exist")
```

However, if you wish to do something with the invitation data (e.g. register the user with the e-mail saved in the invitation), you'll have to implement the register router yourself. You can copy/paste the FastAPI Users one as a basis and customize it, it's quite straightforward:

https://github.com/frankie567/fastapi-users/blob/7ac5dc8486daaaba0fcc58aad074bf33174d603d/fastapi_users/router/register.py#L19-L34

PS: Thank you so much for the sponsorship ❤️ I'll add you to the contributors!

**Reply by rbracco** [reply]
Thank you so much for the detailed reply. I tried implementing #1 but no matter what I do I'm getting back a 422 Unprocessable Entity that goes away if I comment out `dependencies=[Depends(validate_invite)]`. I confirmed via network tab that the request body does contain a string called `invite` but the print statement in `validate_invite` never gets triggered.  I also tried adding invite to the User/UserCreate/UserTable models even though it shouldn't be necessary because I wanted to try everything I could think of to avoid wasting your time. Please let me know if you have any ideas and if not I'll pivot and try solution #2. Cheers.

```python
app.include_router(
    fastapi_users.get_register_router(on_after_register),
    prefix="/auth",
    tags=["auth"],
    dependencies=[Depends(validate_invite)],
)

//validate_invite_code is just checking a json file for valid invites that havent been used, i tried async and non async, 
//I'm just trying to get it wired up and then I'll refactor the actual validation code into this function
async def validate_invite(invite: str = Query(...)):
    print("Validating invite")
    invitation = await validate_invite_code(invite)
    if not invitation:
        raise HTTPException(status_code=403, detail="Invalid Invite Code")

async def validate_invite_code(code):
    invites = read_json(PATH_INVITES_JSON)
    if code not in invites:
        print("Invites code invalid")
    elif invites[code]["used"]:
        print("This code has already been used")
    elif not invites[code]["shared"]:
        print("Hmm this code was never shared")
    else:
        print("this code is valid")
        invites[code]["used"] = True
        write_json(invites, PATH_INVITES_JSON)
        return True
    return False
```

**Reply by frankie567** [reply]
A `Query` parameter shouldn't be in the body, but rather in the URL, as query parameters: `http://localhost:8000/auth/register?invite=XXXXX` : https://fastapi.tiangolo.com/tutorial/query-params-str-validations/?h=query%28#query-parameters-and-string-validations

In this configuration, you can't put it in the request body because it'll be catched by the register route (which expects a valid user registration body).

**Reply by rbracco** [reply]
Wow, seems obvious now haha, not sure how I missed that. Thanks for pointing it out and for all your help.
